### PR TITLE
feat: ML based seed filtering

### PR DESCRIPTION
**Pull request by Corentin-Allaire**
This pull request add a ML based seed selection that can be used to reduce the number of seed after the seeding step.
This PR is pending on the merging of #2690.

### CODE REVIEW

File: Examples/Scripts/Python/MLAmbiguityResolution/seed_filter_full_chain.py [hunk 1]
```diff
@@ -0,0 +1,474 @@
...
+print("Inference : ", (t4 - t3) * 1000 / len(CKF_files), "ms")
+print("Perf : ", (tend - t4) * 1000 / len(CKF_files), "ms")
+print("tot : ", (t4 - start) * 1000 / len(CKF_files), "ms")
+print("Seed filter : ", (t4 - t1) * 1000 / len(CKF_files), "ms")
+
+
+# ==================================================================
+# Plotting
+
+# Combine the events together to have a better statistics
```
Comment (AJPfleger) [inline Examples/Scripts/Python/MLAmbiguityResolution/seed_filter_full_chain.py:322]:
```suggestion
# Combine the events to have a better statistic
```

File: Examples/Scripts/Python/MLAmbiguityResolution/seed_filter_full_chain.py [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+) / len(
+    cleanedDataPlots.loc[cleanedDataPlots["good/duplicate/fake"] == "fake"]["score"]
+)
+plt.hist(
+    cleanedDataPlots.loc[cleanedDataPlots["good/duplicate/fake"] == "fake"]["score"],
+    bins=100,
+    weights=weightsFake,
+    alpha=0.65,
+    label="fake",
+)
```
Comment (AJPfleger) [inline Examples/Scripts/Python/MLAmbiguityResolution/seed_filter_full_chain.py:373]:
```suggestion
for tag in ["good", "duplicate", "fake"]:
        weights = np.ones_like(
            cleanedDataPlots.loc[cleanedDataPlots["good/duplicate/fake"] == tag]["score"]
        ) / len(
            cleanedDataPlots.loc[cleanedDataPlots["good/duplicate/fake"] == tag]["score"]
        )
        plt.hist(
            cleanedDataPlots.loc[cleanedDataPlots["good/duplicate/fake"] == tag]["score"],
            bins=100,
            weights=weights,
            alpha=0.65,
            label=tag,
        )
```

File: Examples/Scripts/Python/MLAmbiguityResolution/seed_solver_network.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+import ast
+
+
+def prepareDataSet(data: pd.DataFrame) -> pd.DataFrame:
+    """Format the dataset that have been written from the Csv file"""
+    """
+    @param[in] data: input DataFrame containing 1 event
+    @return: Formatted DataFrame 
+    """
+    data = data
```
Comment (AJPfleger) [inline Examples/Scripts/Python/MLAmbiguityResolution/seed_solver_network.py:17]:
```suggestion
```

File: Examples/Scripts/Python/MLAmbiguityResolution/train_seed_solver.py [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+from sklearn.preprocessing import LabelEncoder, StandardScaler, OrdinalEncoder
+
+from seed_solver_network import (
+    prepareDataSet,
+    DuplicateClassifier,
+    Normalise,
+)
+
+avg_mean = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
+avg_sdv = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
```
Comment (AJPfleger) [inline Examples/Scripts/Python/MLAmbiguityResolution/train_seed_solver.py:20]:
```suggestion
avg_mean = [0] * 14
avg_sdv = [0] * 14
```

**Review by AJPfleger — CHANGES_REQUESTED** [review]

File: Examples/Algorithms/TrackFindingML/src/SeedFilterMLAlgorithm.cpp [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+  }
+  if (m_cfg.inputSimSeeds.empty()) {
+    throw std::invalid_argument("Missing trajectories input collection");
+  }
+  if (m_cfg.outputTrackParameters.empty()) {
+    throw std::invalid_argument("Missing trajectories output collection");
+  }
+  if (m_cfg.outputSimSeeds.empty()) {
+    throw std::invalid_argument("Missing trajectories output collection");
+  }
```
[the remark this answers is not in the document]
Comment (Corentin-Allaire) [reply Examples/Algorithms/TrackFindingML/src/SeedFilterMLAlgorithm.cpp:34]: No it is not, I copied and paste those line while forgeting to update the text...

File: Examples/Scripts/Python/MLAmbiguityResolution/match_good_track-seed.py [hunk 6]
```diff
@@ -61,7 +59,7 @@ def matchGood(seed_files: list[str], ckf_files: list[str]) -> pd.DataFrame:
         cleanedData = cleanedData.drop(columns=["goodSeed"])
         cleanedData.to_csv(cleaned)
 
-    return data
+    return
```
[the remark this answers is not in the document]
Comment (Corentin-Allaire) [reply Examples/Scripts/Python/MLAmbiguityResolution/match_good_track-seed.py:62]: It does work (python never actually check if the function signature are correct). But I am going to fix this

File: Plugins/Onnx/include/Acts/Plugins/Onnx/SeedClassifier.hpp
[the remark this answers is not in the document]
Comment (Corentin-Allaire) [reply Plugins/Onnx/include/Acts/Plugins/Onnx/SeedClassifier.hpp:1]: For the first point, the "reason" was that the output of the clustering used int. But that is a bad reason and it should actually be a std::size_t. I am changing this. 

As for the ids, I seem to recall that in the geometry layerID, surfaceID and all that are always capitalised

[the remark this answers is not in the document]
Comment (AJPfleger) [reply Plugins/Onnx/include/Acts/Plugins/Onnx/SeedClassifier.hpp:1]: Oh, yes I had a look and `seedID` seems more consistent and only the writer has the off capitalisation.

[the remark this answers is not in the document]
Comment (Corentin-Allaire) [reply Plugins/Onnx/include/Acts/Plugins/Onnx/SeedClassifier.hpp:1]: I might as well fix the spelling in the writer while I am doing this

File: Examples/Algorithms/TrackFindingML/src/SeedFilterMLAlgorithm.cpp [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+    clusteringParams.push_back(
+        {phi / m_cfg.clusteringWeighPhi, eta / m_cfg.clusteringWeighEta,
+         seeds[i].z() / m_cfg.clusteringWeighZ, pT / m_cfg.clusteringWeighPt});
+    // Fill the NN input
+    networkInput.row(NNindex) << pT, eta, phi, seeds[i].sp()[0]->x(),
+        seeds[i].sp()[0]->y(), seeds[i].sp()[0]->z(), seeds[i].sp()[1]->x(),
+        seeds[i].sp()[1]->y(), seeds[i].sp()[1]->z(), seeds[i].sp()[2]->x(),
+        seeds[i].sp()[2]->y(), seeds[i].sp()[2]->z(), seeds[i].z(),
+        seeds[i].seedQuality();
+  }
```
[the remark this answers is not in the document]
Comment (Corentin-Allaire) [reply Examples/Algorithms/TrackFindingML/src/SeedFilterMLAlgorithm.cpp:76]: I think it used to be there in case the indices in the NN and in the collection where different. But now you are right this doesn't do anything !

File: Examples/Scripts/Python/MLAmbiguityResolution/ambiguity_solver_perf.py [hunk 8]
```diff
@@ -106,3 +152,13 @@ def readDataSet(CKS_files: list[str]) -> pd.DataFrame:
...
+
+print("===computed efficiencies ML===")
+print("nb particles : ", nb_part)
+print("nb good match : ", nb_good_match_ML)
+print("nb particle reco : ", nb_reco_part_ML)
+print("nb track reco : ", nb_reco_track_ML)
+print("Efficiency (good track) : ", 100 * nb_good_match_ML / nb_part, " %")
+print("Efficiency (particle reco) : ", 100 * nb_reco_part_ML / nb_part, " %")
+print("duplicate rate: ", 100 * nb_reco_duplicate_ML / nb_reco_track_ML, " %")
+print("Fake rate: ", 100 * nb_reco_fake_ML / nb_reco_track_ML, " %")
```
[the remark this answers is not in the document]
Comment (Corentin-Allaire) [reply Examples/Scripts/Python/MLAmbiguityResolution/ambiguity_solver_perf.py:164]: I always end-up using the French convention on the `:` by accident... I will change that

File: docs/plugins/MLAlgorithms.md [hunk 9]
```diff
@@ -1,38 +1,64 @@
...
 
-Additionally, two other Python files are present with the training. One called `ambiguity_solver_full_chain.py` this one perform the same job as the chain we presented here but in python using Csv files in input. The difference is that they also perform an additional pre-clustering using a DBScan. The other one `ambiguity_solver_perf.py` can be used to study the performance of the ambiguity solver (ML and not) and also takes Csv files as input.
\ No newline at end of file
+Additionally, two other Python files are present with the training. One called `seed_filter_full_chain.py` performs the same job as the chain we presented here, but in Python, using CSV files in the input, this will plot many helpful distributions that will help you understand how the Filter is performing.
```
[the remark this answers is not in the document]
Comment (Corentin-Allaire) [reply docs/plugins/MLAlgorithms.md:64]: I copy paste part of the text from the Solver but the is only one file... I will update that part

File: Examples/Scripts/Python/MLAmbiguityResolution/seed_solver_network.py (hunk 3 above)
Comment (Corentin-Allaire) [reply Examples/Scripts/Python/MLAmbiguityResolution/seed_solver_network.py:17]: I thing there is a reason for that one. I remember having to write this or it wouldn't work... Let me see if I can recall why

Comment (Corentin-Allaire) [reply Examples/Scripts/Python/MLAmbiguityResolution/seed_solver_network.py:18]: Well removing seems to be ok so lets do it

**Review by AJPfleger — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :white_check_mark: Athena integration test results [61bc4e1fa2695d1db2f0e11503b8cea660998e5b]
### :white_check_mark: All tests successful

|status|job|report|
|------|---|------|
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34739749)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsBenchmarkWithSpot.sh 8 100](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34739748)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsWorkflow.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34739747)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateAmbiguityResolution.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34739746)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateResolvedTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34739745)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34739744)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsCoreSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34739743)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34739742)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34739741)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateOrthogonalSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34739740)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateClusters.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34739739)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsPersistifyEDM.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34739738)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsGSFRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34739737)||
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsKfRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34739736)||
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsExtrapolationAlgTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34739735)||
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsITkTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34739734)||
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34739733)||
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34739732)||
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34739731)||
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34739730)||
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34739729)||
|:green_circle: | [run_art_test: test_data18_13TeV_1000evt](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34739728)||
|:green_circle: | [run_art_test: test_ttbarPU40_reco](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/34739727)||
